include/cell.h

```diff
@@ -217,6 +223,28 @@ class Cell {
   //! Return previous mpi rank
   unsigned previous_mpirank() const;
 
+  //! Assign free surface
+  //! \param[in] free_surface boolean indicating free surface cell
+  void assign_free_surface(bool free_surface) { free_surface_ = free_surface; };
+
+  //! Return free surface bool
+  //! \retval free_surface_ indicating free surface cell
+  bool free_surface() const { return free_surface_; };
+
+  //! Assign volume traction to node
+  //! \param[in] volume_fraction cell volume fraction
+  void assign_volume_fraction(double volume_fraction) {
+    volume_fraction_ = volume_fraction;
+  };
+
+  //! Return cell volume fraction
+  //! \retval volume_fraction_ cell volume fraction
+  double volume_fraction() const { return volume_fraction_; };
+
+  //! Map cell volume to the nodes
+  //! \param[in] phase to map volume
+  void map_cell_volume_to_nodes(unsigned phase);
```

Comment:
Would this affect GIMP and other non-local algorithms?

Reply:
Hmm... That's a good point, but I don't think the free-surface detection can be used for non-local MPM yet in general, so I am expecting this function won't work well too for non-local MPM.